File: Sprite.ts
import { GamePanel } from './GamePanel';

export abstract class Sprite {
    x: number;
    y: number;
    image: HTMLImageElement;
    active: boolean = true;

    constructor(img: HTMLImageElement, x: number, y: number) {
        this.image = img;
        this.x = x;
        this.y = y;
    }

    draw(ctx: CanvasRenderingContext2D): void {
        if (this.active && this.image.complete) {
            // draw at the image's natural size
            ctx.drawImage(this.image, this.x, this.y);
        }
    }

    abstract update(panel: GamePanel, collisionEnabled: boolean): void;

    getBounds(): DOMRect {
        // use the actual image dimensions for collisions
        const w = this.image.naturalWidth;
        const h = this.image.naturalHeight;
        return new DOMRect(this.x, this.y, w, h);
    }
}

File: Sprite.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.mazegame;

/**
 *
 * @author lsuti
 */
import java.awt.Graphics2D;
import java.awt.Rectangle;
import java.awt.image.BufferedImage;

public abstract class Sprite {

    protected int x, y;
    protected BufferedImage image;
    protected boolean active = true;
    protected int width, height;

    public Sprite(BufferedImage img, int x, int y) {
        this.image = img;
        this.x = x;
        this.y = y;
        if (img != null) {
            this.width = img.getWidth();
            this.height = img.getHeight();
        }
    }

    public void draw(Graphics2D g) {
        if (active && image != null) {
            g.drawImage(image, x, y, null);
        }
    }

    /**
     * Update your internal state each tick.
     */
    public abstract void update(GamePanel panel, boolean collisionEnabled);

    public Rectangle getBounds() {
        return new Rectangle(x, y, width, height);
    }

    // getters/setters
    public boolean isActive() {
        return active;
    }

    public void setActive(boolean a) {
        this.active = a;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public void setX(int x) {
        this.x = x;
    }

    public void setY(int y) {
        this.y = y;
    }
}
